File: src/go/finished/find_the_length_of_the_longest_common_prefix.go
package leetcode

// https://leetcode.com/problems/find-the-length-of-the-longest-common-prefix/

import "strconv"

func longestCommonPrefix(arr1 []int, arr2 []int) int {
	trie1 := newtrie()
	for _, n1 := range arr1 {
		trie1.insert(strconv.Itoa(n1))
	}

	maxLen := -1

	for _, n2 := range arr2 {
		maxLen = max(
			maxLen,
			trie1.longestPrefix(strconv.Itoa(n2)),
		)
	}

	return maxLen
}

type trie struct {
	isEnd    bool
	children []*trie
}

func newtrie() *trie {
	return &trie{
		children: make([]*trie, 10),
	}
}

func (t *trie) insert(s string) {
	if s == "" {
		t.isEnd = true
		return
	}

	c := int(s[0] - '0')
	rest := s[1:]

	if t.children[c] == nil {
		t.children[c] = newtrie()
	}
	t.children[c].insert(rest)
}

func (t *trie) longestPrefix(s string) int {
	if s == "" {
		return 0
	}

	c := int(s[0] - '0')
	rest := s[1:]

	if t.children[c] == nil {
		return 0
	}

	return 1 + t.children[c].longestPrefix(rest)
}

File: src/python/finished/find_the_length_of_the_longest_common_prefix.py
from typing import List

# https://leetcode.com/problems/find-the-length-of-the-longest-common-prefix


class Trie:
    def __init__(self):
        self.children = {}

    def insert(self, s: str) -> None:
        cur = self
        for i in range(len(s)):
            ch = s[i]
            cur = cur.children.setdefault(ch, Trie())
        return

    def __str__(self):
        builder = ["("]
        for k, v in self.children.items():
            builder.append(k)
            builder.append(":")
            builder.append(str(v))
        builder.append(")")
        return "".join(builder)


class Solution:
    def longestCommonPrefix(self, arr1: List[int], arr2: List[int]) -> int:
        t1, t2 = Trie(), Trie()
        for n in arr1:
            # print("1", n)
            t1.insert(str(n))
        for n in arr2:
            # print("2", n)
            t2.insert(str(n))
        # print(t1)
        # print(t2)

        def search(t1: Trie, t2: Trie) -> int:
            keys = set()
            keys.update(t1.children.keys())
            keys.update(t2.children.keys())

            best = 0
            for k in keys:
                if not (k in t1.children and k in t2.children):
                    continue
                best = max(best, 1 + search(t1.children[k], t2.children[k]))
            return best

        return search(t1, t2)
